pkg/agent/plugin/nodeattestor/k8s/k8s.go

```diff
@@ -0,0 +1,245 @@
+package k8s
+
+import (
+	"context"
+	"crypto"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"sync"
+
+	"github.com/hashicorp/hcl"
+	"github.com/spiffe/spire/pkg/common/plugin/k8s"
+	"github.com/spiffe/spire/pkg/common/plugin/x509pop"
+	"github.com/spiffe/spire/proto/agent/nodeattestor"
+	"github.com/spiffe/spire/proto/common"
+	"github.com/spiffe/spire/proto/common/plugin"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/cert"
+	"k8s.io/client-go/util/certificate/csr"
+)
+
+const (
+	pluginName = "k8s"
+)
+
+type configData struct {
+	spiffeID        string
+	privateKey      crypto.PrivateKey
+	attestationData *common.AttestationData
+}
+
+type K8sConfig struct {
+	TrustDomain     string `hcl:"trust_domain"`
+	PrivateKeyPath  string `hcl:"k8s_private_key_path"`
+	CertificatePath string `hcl:"k8s_certificate_path"`
+	K8sCACertPath   string `hcl:"k8s_ca_certificate_path"`
+	KubeconfigPath  string `hcl:"kubeconfig_path"`
+}
+
+type K8sPlugin struct {
+	m          sync.Mutex
+	c          *K8sConfig
+	kubeClient kubernetes.Interface
+}
+
+var _ nodeattestor.Plugin = (*K8sPlugin)(nil)
+
+func (p *K8sPlugin) getConfig() *K8sConfig {
+	p.m.Lock()
+	defer p.m.Unlock()
+	return p.c
+}
+
+func (p *K8sPlugin) setConfig(c *K8sConfig) {
+	p.m.Lock()
+	defer p.m.Unlock()
+	p.c = c
+}
+
+func getAgentName() string {
+	name, err := os.Hostname()
+	if err != nil {
+		name = "unknown"
+	}
+	return name
+}
+
+func getKubeClient(kubeConfigFilePath, clientCertFilePath, clientKeyFilePath, caCertFilePath string) (kubernetes.Interface, error) {
+	if kubeConfigFilePath == "" {
+		// Try KUBECONFIG env variable
+		kubeConfigFilePath = os.Getenv("KUBECONFIG")
+		if kubeConfigFilePath == "" {
+			// Still no luck, try default (home)
+			home := os.Getenv("HOME")
+			if home != "" {
+				kubeConfigFilePath = path.Join(home, ".kube", "config")
+			}
+		}
+	}
+
+	if kubeConfigFilePath == "" {
+		return nil, fmt.Errorf("Unable to locate kubeconfig")
+	}
+
+	config, err := clientcmd.BuildConfigFromFlags("", kubeConfigFilePath)
+	if err != nil {
+		return nil, fmt.Errorf("Error accessing kubeconfig %s: %v", kubeConfigFilePath, err)
+	}
+
+	config.TLSClientConfig.CertFile = clientCertFilePath
+	config.TLSClientConfig.KeyFile = clientKeyFilePath
+	config.TLSClientConfig.CAFile = caCertFilePath
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, fmt.Errorf("Error creating clientset: %v", err)
+	}
+	return clientset, nil
+}
+
+func fetchK8sCert(kubeClient kubernetes.Interface) (*tls.Certificate, error) {
+	key, err := cert.MakeEllipticPrivateKeyPEM()
+	if err != nil {
+		return nil, fmt.Errorf("Error creating private key: %v", err)
+	}
+
+	certsIntf := kubeClient.CertificatesV1beta1().CertificateSigningRequests()
+	cert, err := csr.RequestNodeCertificate(certsIntf, key, types.NodeName(getAgentName()))
```

Comment:
Does the CSR signing process in kubernetes do anything to assert the correct value in NodeName? In other words, could I get kubernetes to sign a CSR with the name of a different node here?

Reply:
Right now it does not. It just checks that CN conforms to the expected format and key usage is what is expected. There are plans to strengthen this and support proof-of-ownership, TPMs, etc.

Options for improving over the baseline are:
- Disable automatic CSR approval, rely on Kubernetes admin to verify the content of the CSR manually before approving it
- Create a custom CSR signer that will perform extra checks/challenges
- K8s attestor in SPIRE server checks the IP address of the agent connection and compares it against what is in the Kubernetes certificate

I think all options are doable but IMO, unless this is a showstopper, the best course of action is to live with the limitation for the time being and leverage the improvements of the Kubernetes node attestation mechanism as they become available.

Reply:
We discussed this topic on the SIG-Kubernetes call today. The biggest drawback here is that we are essentially asserting the ability of the caller to get a CSR approved, and not asserting anything about the actual identity of the caller. In certain situations, this can result in a non-agent workload being able to obtain the identities of all other workloads in the cluster.

To mitigate this, we should try to assert that the caller is in fact an agent. We discussed using a service account token to do this, configuring the server node attestor with the namespace and service account that the agent runs as. This token can either 1) be used as credentials to submit the CSR, and verified by the server, or 2) passed directly to the server.

Taking this approach gives stronger assurance that the caller we are attesting is an agent, however it still allows for a dishonest agent (or anything that can run as the named service account) to claim that it is a different node/agent. To mitigate this, we could use the kubelet identity to submit the CSR rather than the service account. The server attestor can then verify that the kubelet identity which submitted the CSR matches the SPIFFE ID we are about to issue. IMO, this is the ideal approach, and provides the strongest level of attestation of all the options. @enricoschiattarella has mentioned that obtaining the kubelet identity may be non-trivial, but perhaps it warrants some further research.

To avoid the non-agents-could-obtain-arbitrary-identities problem, we should either add the service account token logic, or use the kubelet identity directly for CSR submission, with a preference for the latter. If we choose the service account path, it probably warrants a direct call out in the plugin documentation (perhaps in the form of a Security Considerations section?). SPIRE is known for its ability to strongly attest and issue identity, but some sacrifices are made in SA token approach (namely, that we are not actually asserting the identity of a node, rather we are asserting that the caller lives in a particular service account).

Some related notes:
* Using SA token for CSR submission instead of sending it directly to the server avoids the need to pass a SA token to a system "outside" of the kubernetes control plane
* @mattmoyer informed us that the identity used to submit the CSR is included in the signed certificate - this could be directly extracted by the server attestor for validation, as opposed to having to hit the api server to interrogate the identity of the submitter
* Node identity may be embedded in SA tokens in the future, it is unclear whether the node identity will 1) be included in resulting signed certificates, 2) be discoverable through interrogation of the API server, looking at the CSR approval stuff. If neither 1 or 2, we would end up having to pass the token directly to SPIRE anyways.